File: esim-frontend/esim-ide/schematic/styletemplate.cc
/* ****************************************************************************
 * OpenESIM (Electrical Engineering Simulator)                                *
 * Copyright (C) 2019, ESIM team. <diyer175@hotmail.com>                      *
 *                                                                            *
 * This project is free software; you can redistribute it and/or              *
 * modify it under the terms of the GNU Lesser General Public License(GPL)    *
 * as published by the Free Software Foundation; either version 2.1           *
 * of the License, or (at your option) any later version.                     *
 *                                                                            *
 * This project is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU          *
 * Lesser General Public License for more details.                            *
 * ****************************************************************************/

#include "styletemplate.h"
#include "persistencedom.h"
#include "errtrace.h"
#include <cassert>

namespace schematic
{

////////////////////////////////////////////////////////////////////////////////
// Line Style Template                                                        //
////////////////////////////////////////////////////////////////////////////////

LineStyle::LineStyle(const std::string &symbol, const templatedColor & color, float width, templatedStyle line) :
    StyleBase(symbol),
    m_color(color),
    m_width(width),
    m_line(line)
{
  set();
}

LineStyle::LineStyle(const std::string &symbol, const QPen &penIns) :
    StyleBase(symbol),
    m_color(penIns.color()),
    m_width(penIns.widthF()),
    m_penIns(penIns)
{
    switch(penIns.style())
    {
      case Qt::NoPen:
        m_line = LINE_NONE; break;
      case Qt::SolidLine:
        m_line = LINE_SOLID; break;
      case Qt::DashLine:
        m_line = LINE_DASH; break;
      case Qt::DotLine:
        m_line = LINE_DOT; break;
      case Qt::DashDotLine:
        m_line = LINE_DASH_DOT; break;
      case Qt::DashDotDotLine:
        m_line = LINE_DASH_DOTDOT; break;
      default:
        m_line = LINE_SOLID;
    }
}

void LineStyle::set()
{
  Qt::PenStyle attr = Qt::SolidLine;
  switch(m_line)
  {
    case LINE_NONE:
      attr = Qt::NoPen; break;
    case LINE_SOLID:
      attr = Qt::SolidLine; break;
    case LINE_DASH:
      attr = Qt::DashLine; break;
    case LINE_DOT:
      attr = Qt::DotLine; break;
    case LINE_DASH_DOT:
      attr = Qt::DashDotLine; break;
    case LINE_DASH_DOTDOT:
      attr = Qt::DashDotDotLine; break;
    default:
      attr = Qt::SolidLine;
  }
  m_penIns = QPen(QColor(m_color.R, m_color.G, m_color.B), m_width, attr);
}

int LineStyle::serialize(persistence::PersistenceDom *dom)
{
  bool custom;
  RC_CHECK_RETURN(dom->ser(m_symbol));

  const LineStyle* ref = StyleTemplates::getInstance()->getLineStyleRef(m_symbol);

  custom = !(this->m_color == ref->m_color);
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    {
      RC_CHECK_RETURN(dom->ser(m_color.R));
      RC_CHECK_RETURN(dom->ser(m_color.G));
      RC_CHECK_RETURN(dom->ser(m_color.B));
    }
    
  custom = this->m_width != ref->m_width;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser(m_width)); }

  custom = this->m_line != ref->m_line;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser((int)m_line)); }

  return 0;
}

int LineStyle::deserialize(persistence::PersistenceDom *dom)
{
  bool custom;
  RC_CHECK_RETURN(dom->des(&m_symbol));

  const LineStyle* ref = StyleTemplates::getInstance()->getLineStyleRef(m_symbol);

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    {
      RC_CHECK_RETURN(dom->des(&m_color.R));
      RC_CHECK_RETURN(dom->des(&m_color.G));
      RC_CHECK_RETURN(dom->des(&m_color.B));
    }
  else
    {
      m_color = ref->m_color;
    }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    { RC_CHECK_RETURN(dom->des(&m_width)); }
  else
    { m_width = ref->m_width; }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    {
      int line;
      RC_CHECK_RETURN(dom->des(&line));
      m_line = static_cast<templatedStyle>(line);
    }
  else
    {
      m_line = ref->m_line;
    }
  set();
  return 0;
}

bool LineStyle::operator ==(const LineStyle &src) const
{
  return (m_color == src.m_color &&
      m_width == src.m_width &&
      m_line == src.m_line);
}

////////////////////////////////////////////////////////////////////////////////
// Fill Style Template                                                        //
////////////////////////////////////////////////////////////////////////////////

FillStyle::FillStyle(const std::string &symbol,
                     const templatedColor &color,
                     const templatedColor &bkcolor,
                     float width,
                     templatedStyle line,
                     templatedStyle brush)
  : LineStyle(symbol, color, width, line),
    m_bkcolor(bkcolor),
    m_brush(brush)
{
  set();
}

FillStyle::FillStyle(const std::string &symbol, const QPen &penIns, const QBrush &brushIns)
  : LineStyle(symbol, penIns),
    m_bkcolor(brushIns.color()),
    m_brushIns(brushIns)
{
  switch(brushIns.style())
  {
    case Qt::NoBrush:
      m_brush = BRUSH_NONE; break;
    case Qt::SolidPattern:
      m_brush = BRUSH_SOLID; break;
    case Qt::HorPattern:
      m_brush = BRUSH_HOR; break;
    case Qt::VerPattern:
      m_brush = BRUSH_VER; break;
    case Qt::CrossPattern:
      m_brush = BRUSH_CROSS; break;
    default:
      m_brush = BRUSH_NONE;
  }
}

void FillStyle::set()
{
  LineStyle::set();
  Qt::BrushStyle attr = Qt::NoBrush;
  switch(m_brush)
  {
    case BRUSH_NONE:
      attr = Qt::NoBrush; break;
    case BRUSH_SOLID:
      attr = Qt::SolidPattern; break;
    case BRUSH_HOR:
      attr = Qt::HorPattern; break;
    case BRUSH_VER:
      attr = Qt::VerPattern; break;
    case BRUSH_CROSS:
      attr = Qt::CrossPattern; break;
    default:
      attr = Qt::NoBrush;
  }

  m_brushIns = QBrush(QColor(m_bkcolor.R,m_bkcolor.G,m_bkcolor.B), attr);
}

int FillStyle::serialize(persistence::PersistenceDom *dom)
{
  bool custom;
  RC_CHECK_RETURN(dom->ser(m_symbol));

  const FillStyle* ref = StyleTemplates::getInstance()->getFillStyleRef(m_symbol);

  custom = !(this->m_color == ref->m_color);
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    {
      RC_CHECK_RETURN(dom->ser(m_color.R));
      RC_CHECK_RETURN(dom->ser(m_color.G));
      RC_CHECK_RETURN(dom->ser(m_color.B));
    }
  custom = !(this->m_bkcolor == ref->m_bkcolor);
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    {
      RC_CHECK_RETURN(dom->ser(m_bkcolor.R));
      RC_CHECK_RETURN(dom->ser(m_bkcolor.G));
      RC_CHECK_RETURN(dom->ser(m_bkcolor.B));
    }

  custom = this->m_width != ref->m_width;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser(m_width)); }

  custom = this->m_line != ref->m_line;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser((int)m_line)); }

  custom = this->m_brush != ref->m_brush;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser((int)m_brush)); }
  return 0;
}

int FillStyle::deserialize(persistence::PersistenceDom *dom)
{
  bool custom;
  RC_CHECK_RETURN(dom->des(&m_symbol));

  const FillStyle* ref = StyleTemplates::getInstance()->getFillStyleRef(m_symbol);

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    {
      RC_CHECK_RETURN(dom->des(&m_color.R));
      RC_CHECK_RETURN(dom->des(&m_color.G));
      RC_CHECK_RETURN(dom->des(&m_color.B));
    }
  else
    {
      m_color = ref->m_color;
    }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    {
      RC_CHECK_RETURN(dom->des(&m_bkcolor.R));
      RC_CHECK_RETURN(dom->des(&m_bkcolor.G));
      RC_CHECK_RETURN(dom->des(&m_bkcolor.B));
    }
  else
    {
      m_bkcolor = ref->m_bkcolor;
    }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    { RC_CHECK_RETURN(dom->des(&m_width)); }
  else
    { m_width = ref->m_width; }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    {
      int sty;
      RC_CHECK_RETURN(dom->des(&sty));
      m_line = static_cast<templatedStyle>(sty);
    }
  else
    {
      m_line = ref->m_line;
    }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    {
      int sty;
      RC_CHECK_RETURN(dom->des(&sty));
      m_brush = static_cast<templatedStyle>(sty);
    }
  else
    {
      m_brush = ref->m_brush;
    }
  set();
  return 0;
}

bool FillStyle::operator ==(const FillStyle &src) const
{
  return (LineStyle::operator ==(src) &&
      (m_bkcolor == src.m_bkcolor &&
      m_brush == src.m_brush));
}

////////////////////////////////////////////////////////////////////////////////
// Text Style Template                                                        //
////////////////////////////////////////////////////////////////////////////////

TextStyle::TextStyle(const std::string &symbol,
                     const templatedColor & color,
                     const std::string &fontName,
                     int sizePt,
                     bool bold,
                     bool italic,
                     bool underline) :
    StyleBase(symbol),
    m_color(color),
    m_fontName(fontName),
    m_sizePt(sizePt),
    m_bold(bold),
    m_italic(italic),
    m_underline(underline)
{
  set();
}

TextStyle::TextStyle(const std::string &symbol, const QPen &penIns, const QFont &fontIns) :
    StyleBase(symbol),
    m_color(penIns.color()),
    m_fontName(fontIns.family().toStdString()),
    m_sizePt(fontIns.pointSize()),
    m_bold(fontIns.bold()),
    m_italic(fontIns.italic()),
    m_underline(fontIns.underline()),
    m_penIns(penIns),
    m_fontIns(fontIns)
{
}

void TextStyle::set()
{
  m_fontIns = QFont(m_fontName.c_str(), m_sizePt);
  m_fontIns.setBold(m_bold);
  m_fontIns.setItalic(m_italic);
  m_fontIns.setUnderline(m_underline);
  m_penIns = QPen(QColor(m_color.R, m_color.G, m_color.B));
}

int TextStyle::serialize(persistence::PersistenceDom *dom)
{
  bool custom;
  RC_CHECK_RETURN(dom->ser(m_symbol));

  const TextStyle* ref = StyleTemplates::getInstance()->getTextStyleRef(m_symbol);

  custom = !(this->m_color == ref->m_color);
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    {
      RC_CHECK_RETURN(dom->ser(m_color.R));
      RC_CHECK_RETURN(dom->ser(m_color.G));
      RC_CHECK_RETURN(dom->ser(m_color.B));
    }

  custom = this->m_fontName.compare(ref->m_fontName) != 0;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser(m_fontName)); }

  custom = this->m_sizePt != ref->m_sizePt;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser(m_sizePt)); }

  custom = this->m_bold != ref->m_bold;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser(m_bold)); }

  custom = this->m_italic != ref->m_italic;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser(m_italic)); }

  custom = this->m_underline != ref->m_underline;
  RC_CHECK_RETURN(dom->ser(custom));
  if (custom)
    { RC_CHECK_RETURN(dom->ser(m_underline)); }
  return 0;
}

int TextStyle::deserialize(persistence::PersistenceDom *dom)
{
  bool custom;
  RC_CHECK_RETURN(dom->des(&m_symbol));

  const TextStyle* ref = StyleTemplates::getInstance()->getTextStyleRef(m_symbol);

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    {
      RC_CHECK_RETURN(dom->des(&m_color.R));
      RC_CHECK_RETURN(dom->des(&m_color.G));
      RC_CHECK_RETURN(dom->des(&m_color.B));
    }
  else
    {
      m_color = ref->m_color;
    }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    { RC_CHECK_RETURN(dom->des(&m_fontName)); }
  else
    { m_fontName = ref->m_fontName; }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    { RC_CHECK_RETURN(dom->des(&m_sizePt)); }
  else
    { m_sizePt = ref->m_sizePt; }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    { RC_CHECK_RETURN(dom->des(&m_bold)); }
  else
    { m_bold = ref->m_bold; }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    { RC_CHECK_RETURN(dom->des(&m_italic)); }
  else
    { m_italic = ref->m_italic; }

  RC_CHECK_RETURN(dom->des(&custom));
  if (custom)
    { RC_CHECK_RETURN(dom->des(&m_underline)); }
  else
    { m_underline = ref->m_underline; }
  set();
  return 0;
}

bool TextStyle::operator==(const TextStyle &src) const
{
  return (m_color == src.m_color &&
      m_fontName == src.m_fontName &&
      m_sizePt == src.m_sizePt &&
      m_bold == src.m_bold &&
      m_italic == src.m_italic &&
      m_underline == src.m_underline);
}

////////////////////////////////////////////////////////////////////////////////
// Style Templates                                                            //
////////////////////////////////////////////////////////////////////////////////

StyleTemplates *StyleTemplates::m_instance = 0l;

StyleTemplates::StyleTemplates()
{
  m_lineStyles["none"]      = new LineStyle("none", templatedColor(0, 0, 0), 1, LINE_SOLID);
  m_lineStyles["wire"]      = new LineStyle("wire", templatedColor(0, 128, 255), 1, LINE_SOLID);
  m_lineStyles["component"] = new LineStyle("component", templatedColor(128, 0, 0), 1, LINE_SOLID);
  m_lineStyles["pin"]       = new LineStyle("pin", templatedColor(128, 0, 0), 1, LINE_SOLID);

  m_fillStyles["none"]      = new FillStyle("none", templatedColor(0, 0, 0), templatedColor(0, 0, 0), 1, LINE_SOLID, BRUSH_NONE);
  m_fillStyles["component"] = new FillStyle("component", templatedColor(128, 0, 0), templatedColor(0, 0, 0), 1, LINE_SOLID, BRUSH_NONE);

  m_textStyles["none"]      = new TextStyle("none", templatedColor(0, 0, 0), "Times New Roman", 8, false, false, false);
  m_textStyles["component"] = new TextStyle("component", templatedColor(128, 0, 0), "Times New Roman", 8, false, false, false);
  m_textStyles["pin"]       = new TextStyle("pin", templatedColor(128, 0, 0), "Times New Roman", 8, false, false, false);
  m_textStyles["symbol"]    = new TextStyle("symbol", templatedColor(0, 132, 132), "Times New Roman", 8, true, false, false);
}

StyleTemplates::~StyleTemplates()
{
  for(std::map<std::string, LineStyle *>::iterator it=m_lineStyles.begin(); it != m_lineStyles.end(); it++)
    {delete it->second;}
  for(std::map<std::string, FillStyle *>::iterator it=m_fillStyles.begin(); it != m_fillStyles.end(); it++)
    {delete it->second;}
  for(std::map<std::string, TextStyle *>::iterator it=m_textStyles.begin(); it != m_textStyles.end(); it++)
    {delete it->second;}
}

const LineStyle *StyleTemplates::getLineStyleRef(const std::string &symbol)
{
  const LineStyle *line = m_lineStyles[symbol];
  return line ? line : m_lineStyles["none"];
}
const FillStyle *StyleTemplates::getFillStyleRef(const std::string &symbol)
{
  const FillStyle *fill = m_fillStyles[symbol];
  return fill ? fill : m_fillStyles["none"];
}
const TextStyle *StyleTemplates::getTextStyleRef(const std::string &symbol)
{
  const TextStyle *text = m_textStyles[symbol];
  return text ? text : m_textStyles["none"];
}

LineStyle StyleTemplates::copyLineStyle(const std::string &symbol)
{return *getLineStyleRef(symbol);}
FillStyle StyleTemplates::copyFillStyle(const std::string &symbol)
{return *getFillStyleRef(symbol);}
TextStyle StyleTemplates::copyTextStyle(const std::string &symbol)
{return *getTextStyleRef(symbol);}

const char *StyleTemplates::getTemplatedStyleName(templatedStyle style)
{
  switch (style)
  {
    case LINE_NONE:
    case BRUSH_NONE:
      return "None";
    case LINE_SOLID:
    case BRUSH_SOLID:
      return "Solid";
    case LINE_DASH:
      return "Dash";
    case LINE_DOT:
      return "Dot";
    case LINE_DASH_DOT:
      return "Dash Dot";
    case LINE_DASH_DOTDOT:
      return "Dash Dot dot";
    case BRUSH_HOR:
      return "Horizon";
    case BRUSH_VER:
      return "Vertical";
    case BRUSH_CROSS:
      return "Cross";
    default:
      assert(0);
      return "Unknown";
  }
}


} // namespace schematic

File: esim-frontend/test/schematic/styleTemplate.py
#
#  OpenESIM (Electrical Engineering Simulator)
#  Copyright (C) 2019, ESIM team. <diyer175@hotmail.com>
#
#  This project is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public License(GPL)
#  as published by the Free Software Foundation; either version 2.1
#  of the License, or (at your option) any later version.
#
#  This project is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#

from PyQt5.QtCore import Qt
from PyQt5.QtGui import QColor, QPainter, QPen, QBrush, QFont
from enum import IntEnum, unique
from util.persistence import save, get
import copy

@unique
class Style(IntEnum):
    LINE_NONE = 0
    LINE_SOLID = 1
    LINE_DASH = 2
    LINE_DOT = 3
    LINE_DASH_DOT = 4
    LINE_DASH_DOTDOT = 5

    BRUSH_NONE = 10
    BRUSH_SOLID = 11
    BRUSH_HOR = 12
    BRUSH_VER = 13
    BRUSH_CROSS = 14

class LineStyle():
    def __init__(self, styleName='', color=QColor(0,0,0), width=0, line=Style.LINE_NONE):
        self.styleName = styleName
        self.set(color, width, line)
        
    def set(self, color, width, line):
        self.color = color
        self.width = width
        self.line = line
        
        if self.line == Style.LINE_NONE:
            attr = Qt.NoPen
        elif self.line == Style.LINE_SOLID:
            attr = Qt.SolidLine
        elif self.line == Style.LINE_DASH:
            attr = Qt.DashLine
        elif self.line == Style.LINE_DOT:
            attr = Qt.DotLine
        elif self.line == Style.LINE_DASH_DOT:
            attr = Qt.DashDotLine
        elif self.line == Style.LINE_DASH_DOTDOT:
            attr = Qt.DashDotDotLine
        else:
            attr = Qt.SolidLine
        self.penI = QPen(self.color, self.width, attr)
        
    def pen(self):
        return self.penI
        
    def serialize(self, node):
        cust = (self != RefStyle(lineStyles, self.styleName))
        save(node, self.styleName)
        save(node, cust)
        if cust:
            save(node, [self.color.red(), self.color.green(), self.color.blue(),
                        self.width,
                        int(self.line) ])
    
    def deserialize(self, node):
        self.styleName = get(node)
        self.cust = get(node)
        if self.cust:
            red, green, blue, width, line = get(node)
            self.set(QColor(red, green, blue), width, Style(line))
        else:
            src = RefStyle(lineStyles, self.styleName) # use default
            self.set(src.color, src.width, src.line)
            
    def __eq__(self, src):
        return (self.color == src.color and
                self.width == src.width and
                self.line == src.line)
                
class FillStyle(LineStyle):
    def __init__(self, styleName='', color=QColor(0,0,0), bkcolor=QColor(0,0,0), width=0, line=Style.LINE_NONE, brusht=Style.BRUSH_NONE):
        self.styleName = styleName
        self.set(color, bkcolor, width, line, brusht)
        
    # @Override
    def set(self, color, bkcolor, width, line, brusht):
        super().set(color, width, line)
        self.bkcolor = bkcolor
        self.brusht = brusht
        
        if self.brusht == Style.BRUSH_NONE:
            attr = Qt.NoBrush
        elif self.brusht == Style.BRUSH_SOLID:
            attr = Qt.SolidPattern;
        elif self.brusht == Style.BRUSH_HOR:
            attr = Qt.HorPattern;
        elif self.brusht == Style.BRUSH_VER:
            attr = Qt.VerPattern;
        elif self.brusht == Style.BRUSH_CROSS:
            attr = Qt.CrossPattern;
        else:
            attr = Qt.NoBrush;
        self.brushI = QBrush(self.bkcolor, attr)
        
    def brush(self):
        return self.brushI

    # @Override
    def serialize(self, node):
        cust = (self != RefStyle(fillStyles, self.styleName))
        save(node, self.styleName)
        save(node, cust)
        if cust:
            save(node, [self.color.red(), self.color.blue(), self.color.green(),
                        self.bkcolor.red(),self.bkcolor.blue(), self.bkcolor.green(),
                        self.width,
                        int(self.line),
                        int(self.brusht)])

    # @Override
    def deserialize(self, node):
        self.styleName = get(node)
        self.cust = get(node)
        if self.cust:
            red, green, blue, bk_red, bk_green, bk_blue, width, line, brusht = get(node)
            self.set(QColor(red, green, blue), QColor(bk_red, bk_green, bk_blue), width, Style(line), Style(brusht))
        else:
            src = RefStyle(fillStyles, self.styleName) # use default
            self.set(src.color, src.bkcolor, src.width, src.line, src.brusht)

    # @Override
    def __eq__(self, src):
        return (super().__eq__(src) and
                (self.bkcolor == src.bkcolor and
                self.brusht == src.brusht))
            
class TextStyle():
    def __init__(self, styleName='', color=QColor(0,0,0), fontName='', size=0, bold=False, italic=False, underline=False):
        self.styleName = styleName
        self.set(color, fontName, size, bold, italic, underline)
        
    def set(self, color, fontName, size, bold, italic, underline):
        self.color = color
        self.fontName = fontName
        self.size = size
        self.bold = bold
        self.italic = italic
        self.underline = underline
        self.fontI = QFont(fontName, size)
        self.fontI.setBold(bold)
        self.fontI.setItalic(italic)
        self.fontI.setUnderline(underline)
        self.penI = QPen(self.color)
        
    def pen(self):
        return self.penI
    def font(self):
        return self.fontI
        
    def serialize(self, node):
        cust = (self != RefStyle(textStyles, self.styleName))
        save(node, self.styleName)
        save(node, cust)
        if cust:
            save(node, [self.color.red(), self.color.green(), self.color.blue(),
                        self.fontName,
                        self.size,
                        self.bold,
                        self.italic,
                        self.underline])
                    
    def deserialize(self, node):
        self.styleName = get(node)
        cust = get(node)
        if cust:
            red, green, blue, fontName, size, bold, italic, underline = get(node)
            self.set(QColor(red, green, blue), fontName, size, bold, italic, underline)
        else:
            src = RefStyle(textStyles, self.styleName) # use default
            self.set(src.color, src.fontName, src.size, src.bold, src.italic, src.underline)

    def __eq__(self, src):
        return (self.color == src.color and
                self.fontName == src.fontName and
                self.size == src.size and
                self.bold == src.bold and
                self.italic == src.italic and
                self.underline == src.underline)


lineStyles = {}
lineStyles['none']      = LineStyle('none', QColor(0, 0, 0), 1, Style.LINE_SOLID)
lineStyles['wire']      = LineStyle('wire', QColor(0, 128, 255), 1, Style.LINE_SOLID)
lineStyles['component'] = LineStyle('component', QColor(128, 0, 0), 1, Style.LINE_SOLID)
lineStyles['pin']       = LineStyle('pin', QColor(128, 0, 0), 1, Style.LINE_SOLID)

fillStyles = {}
fillStyles['none']      = FillStyle('none', QColor(0, 0, 0), QColor(0, 0, 0), 1, Style.LINE_SOLID, Style.BRUSH_NONE)
fillStyles['component'] = FillStyle('component', QColor(128, 0, 0), QColor(0, 0, 0), 1, Style.LINE_SOLID, Style.BRUSH_NONE)

textStyles = {}
textStyles['none']      = TextStyle('none', QColor(0, 0, 0), 'Times New Roman', 8, False, False, False)
textStyles['component'] = TextStyle('component', QColor(128, 0, 0), 'Times New Roman', 8, False, False, False)
textStyles['pin']       = TextStyle('pin', QColor(128, 0, 0), 'Times New Roman', 8, False, False, False)
textStyles['symbol']       = TextStyle('symbol', QColor(0, 132, 132), 'Times New Roman', 8, True, False, False)


def RefStyle(entry, styleName):
    style = entry.get(styleName)
    if style is not None:
        return style
    return entry['none']
 
def CreateLineStyle(styleName):
    return copy.copy(RefStyle(lineStyles, styleName))
def CreateFillStyle(styleName):
    return copy.copy(RefStyle(fillStyles, styleName))
def CreateTextStyle(styleName):
    return copy.copy(RefStyle(textStyles, styleName))
